### PR TITLE
Apply hardening measures in bitcoind systemd service file

### PR DESCRIPTION
**Pull request by Flowdalic**
Adds typical systemd hardening measurements for network services.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Concept ACK, seems reasonable.

**Comment by fanquake** [comment]
Here's documentation for the 4 new options being added, taken from the [freedesktop.org docs](https://www.freedesktop.org/software/systemd/man/systemd.exec.html).

#### ProtectSystem
Takes a boolean argument or the special values "full" or "strict". If true, mounts the /usr and /boot directories read-only for processes invoked by this unit. If set to "full", the /etc directory is mounted read-only, too. If set to "strict" the entire file system hierarchy is mounted read-only, except for the API file system subtrees /dev, /proc and /sys (protect these directories using PrivateDevices=, ProtectKernelTunables=, ProtectControlGroups=). This setting ensures that any modification of the vendor-supplied operating system (and optionally its configuration, and local mounts) is prohibited for the service. It is recommended to enable this setting for all long-running services, unless they are involved with system updates or need to modify the operating system in other ways. If this option is used, ReadWritePaths= may be used to exclude specific directories from being made read-only. This setting is implied if DynamicUser= is set. For this setting the same restrictions regarding mount propagation and privileges apply as for ReadOnlyPaths= and related calls, see below. Defaults to off.

#### NoNewPrivileges
Takes a boolean argument. If true, ensures that the service process and all its children can never gain new privileges through execve() (e.g. via setuid or setgid bits, or filesystem capabilities). This is the simplest and most effective way to ensure that a process and its children can never elevate privileges again. Defaults to false, but certain settings override this and ignore the value of this setting. This is the case when SystemCallFilter=, SystemCallArchitectures=, RestrictAddressFamilies=, RestrictNamespaces=, PrivateDevices=, ProtectKernelTunables=, ProtectKernelModules=, MemoryDenyWriteExecute=, RestrictRealtime=, or LockPersonality= are specified. Note that even if this setting is overridden by them, systemctl show shows the original value of this setting. Also see No New Privileges Flag.

#### PrivateDevices
Takes a boolean argument. If true, sets up a new /dev mount for the executed processes and only adds API pseudo devices such as /dev/null, /dev/zero or /dev/random (as well as the pseudo TTY subsystem) to it, but no physical devices such as /dev/sda, system memory /dev/mem, system ports /dev/port and others. This is useful to securely turn off physical device access by the executed process. Defaults to false. Enabling this option will install a system call filter to block low-level I/O system calls that are grouped in the @raw-io set, will also remove CAP_MKNOD and CAP_SYS_RAWIO from the capability bounding set for the unit (see above), and set DevicePolicy=closed (see systemd.resource-control(5) for details). Note that using this setting will disconnect propagation of mounts from the service to the host (propagation in the opposite direction continues to work). This means that this setting may not be used for services which shall be able to install mount points in the main mount namespace. The new /dev will be mounted read-only and 'noexec'. The latter may break old programs which try to set up executable memory by using mmap(2) of /dev/zero instead of using MAP_ANON. For this setting the same restrictions regarding mount propagation and privileges apply as for ReadOnlyPaths= and related calls, see above. If turned on and if running in user mode, or in system mode, but without the CAP_SYS_ADMIN capability (e.g. setting User=), NoNewPrivileges=yes is implied.

Note that the implementation of this setting might be impossible (for example if mount namespaces are not available), and the unit should be written in a way that does not solely rely on this setting for security.

#### MemoryDenyWriteExecute
Takes a boolean argument. If set, attempts to create memory mappings that are writable and executable at the same time, or to change existing memory mappings to become executable, or mapping shared memory segments as executable are prohibited. Specifically, a system call filter is added that rejects mmap(2) system calls with both PROT_EXEC and PROT_WRITE set, mprotect(2) or pkey_mprotect(2) system calls with PROT_EXEC set and shmat(2) system calls with SHM_EXEC set. Note that this option is incompatible with programs and libraries that generate program code dynamically at runtime, including JIT execution engines, executable stacks, and code "trampoline" feature of various C compilers. This option improves service security, as it makes harder for software exploits to change running code dynamically. Note that this feature is fully available on x86-64, and partially on x86. Specifically, the shmat() protection is not available on x86. Note that on systems supporting multiple ABIs (such as x86/x86-64) it is recommended to turn off alternative ABIs for services, so that they cannot be used to circumvent the restrictions of this option. Specifically, it is recommended to combine this option with SystemCallArchitectures=native or similar. If running in user mode, or in system mode, but without the CAP_SYS_ADMIN capability (e.g. setting User=), NoNewPrivileges=yes is implied.

**Comment by dongcarl** [comment]
Hi all, I'm working on #12255 which also changes the systemd service file and sets up the pidfile, datadir, and conffile with the right permissions and at the right places. @laanwj told me to coordinate my changes with people here. From what I see things should be compatible except I make the assumption that the service is running as `bitcoin:bitcoin`, which I think we should be according to init.md right?

**Comment by practicalswift** [comment]
Concept ACK. Hardening is good!

File: contrib/init/bitcoind.service [hunk 1]
```diff
@@ -19,7 +19,26 @@ User=bitcoin
 Type=forking
 PIDFile=/run/bitcoind/bitcoind.pid
 Restart=on-failure
+
+# Hardening measurements
```
Comment (laanwj) [inline contrib/init/bitcoind.service]: Nit: Hardening measures, not measurements :)

Comment (Flowdalic) [reply contrib/init/bitcoind.service]: Thanks for pointing this out. Fixed with 8748949.

**Comment by laanwj** [comment]
ACK after squash

**Comment by cdecker** [comment]
ACK :+1:

**Comment by Flowdalic** [comment]
Rebased and squashed.

**Comment by fanquake** [comment]
utACK 2e86249

File: contrib/init/bitcoind.service [hunk 2]
```diff
@@ -19,7 +19,26 @@ User=bitcoin
...
 PrivateTmp=true
 
+# Mount /usr, /boot/ and /etc read-only for the process.
+ProtectSystem=full
+
+# Disallow the process and all of its children to gain
+# new privileges through execve()
+NoNewPrivileges=true
+
+# Use a new /dev namespace only populated with API pseudeo devices
```
Comment (practicalswift) [inline contrib/init/bitcoind.service]: Typo: "pseudo" :-)

File: contrib/init/bitcoind.service [hunk 3]
```diff
@@ -19,7 +19,26 @@ User=bitcoin
...
+# such as /dev/null, /dev/zero or /dev/random
```
Comment (practicalswift) [inline contrib/init/bitcoind.service]: Nit: "or" should be "and"?

File: contrib/init/bitcoind.service [hunk 4]
```diff
@@ -19,7 +19,26 @@ User=bitcoin
...
+PrivateDevices=true
+
+# Deny the creation of writeable and executably memory mappings
```
Comment (practicalswift) [inline contrib/init/bitcoind.service]: Typo: "writable" and "executable" :-)

File: contrib/init/bitcoind.service [hunk 5]
```diff
@@ -19,7 +19,26 @@ User=bitcoin
...
 PIDFile=/run/bitcoind/bitcoind.pid
 Restart=on-failure
+
+# Hardening measures
+####################
+
+# Provide a private /tmp and /var/tmp
 PrivateTmp=true
 
+# Mount /usr, /boot/ and /etc read-only for the process.
```
Comment (practicalswift) [inline contrib/init/bitcoind.service]: Nit: Remove trailing "." to get consistency with the remaining comments.

**Comment by practicalswift** [comment]
utACK modulo review feedback

**Comment by Flowdalic** [comment]
Incorporated @practicalswift's feedback and squashed. Thanks for the feedback. :)

**Comment by fanquake** [comment]
re-utACK 79ddfad

**Pull request merged**